clusterloader2/pkg/imagepreload/imagepreload.go

```diff
@@ -0,0 +1,161 @@
+)
+
+var images = pflag.StringSlice("node-preload-images", []string{}, "List of images to preload on each node in the test cluster before executing tests")
+
+type controller struct {
+	config          *config.ClusterLoaderConfig
+	framework       *framework.Framework
+	templateMapping map[string]interface{}
+	images          []string
+}
+
+// Setup ensures every node in cluster preloads given list of images before starting tests.
+// It does it by creating a Daemonset that call "docker pull" and awaits for Node object to be updated.
+// As a side-effect of the image preloading, size of Node objects is increased.
+//
+// Preloading is skipped in kubemark or if no images have been specified.
+func Setup(conf *config.ClusterLoaderConfig, f *framework.Framework) error {
+	mapping, err := config.GetMapping(conf)
+	if err != nil {
+		return err
+	}
+
+	ctl := &controller{
+		config:          conf,
+		framework:       f,
+		templateMapping: mapping,
+		images:          *images,
+	}
+	return ctl.PreloadImages()
+}
+
+func (i *controller) PreloadImages() error {
+	if len(*images) == 0 {
+		klog.Warning("No images specified. Skipping image preloading")
+		return nil
+	}
+	if i.config.ClusterConfig.Provider == "kubemark" {
+		klog.Warning("Image preloading is disabled in kubemark")
+		return nil
+	}
+
+	clusterSize := i.config.ClusterConfig.Nodes
+	kclient := i.framework.GetClientSets().GetClient()
+
+	// 1. setup node watcher
+	doneNodes := make(map[string]bool)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	ni := informer.NewInformer(
+		kclient,
+		"nodes",
+		util.NewObjectSelector(),
+		func(old, new interface{}) { checkNode(new, doneNodes) })
+	if err := informer.StartAndSync(ni, stopCh, informerTimeout); err != nil {
+		return err
+	}
+
+	// 2. start daemonset
+	klog.Infof("Creating namespace %s...", namespace)
+	if err := client.CreateNamespace(kclient, namespace); err != nil {
+		return err
+	}
+
+	klog.Info("Creating daemonset to preload images...")
+	i.templateMapping["Images"] = i.images
+	if err := i.framework.ApplyTemplatedManifests(manifest, i.templateMapping); err != nil {
+		return err
+	}
+
+	// 3. wait for nodes to contain images
+	klog.Info("Waiting for Node objects to be updated...")
+	if err := wait.Poll(pollingInterval, pollingTimeout, func() (bool, error) {
+		klog.Infof("%d out of %d nodes have pulled images", len(doneNodes), clusterSize)
+		return len(doneNodes) == clusterSize, nil
+	}); err != nil {
+		return nil
+	}
+	klog.Info("Waiting... done")
+
+	// 4. delete daemonset
+	klog.Infof("Deleting namespace %s...", namespace)
+	if err := client.DeleteNamespace(kclient, namespace); err != nil {
+		return err
+	}
+	if err := client.WaitForDeleteNamespace(kclient, namespace); err != nil {
+		return err
+	}
+	return nil
+}
+
+func checkNode(obj interface{}, done map[string]bool) {
+	if obj == nil {
+		return
+	}
+
+	node := obj.(*v1.Node)
+	if hasPreloadedImages(node) {
+		done[node.Name] = true
```

Comment:
Is it a documented and guaranteed feature of an informer that event handlers are called from a single goroutine and hence don't have to be safe thread? If not, and it's only an implementation detail, then I think adding a mutex here wouldn't hurt.

Reply:
We should handle it with mutex. Please add a thread-safe structure to keep this info.

Reply:
Done.